plugin/plugin.go

```diff
@@ -365,3 +365,74 @@ func (p *Plugin) getMessage(t string) *descriptor.DescriptorProto {
 	return file.GetMessage(strings.TrimPrefix(t, "."+file.GetPackage()+"."))
 }
 
+func (p *Plugin) GetDeclinedMethods(options []av_opts.AtlasValidateFieldOption_Operation) []string {
+	httpMethods := make(map[string]struct{}, 0)
+	for _, op := range options {
+		switch op {
+		case av_opts.AtlasValidateFieldOption_create:
+			httpMethods["POST"] = struct{}{}
+		case av_opts.AtlasValidateFieldOption_update:
+			httpMethods["PUT"] = struct{}{}
+			httpMethods["PATCH"] = struct{}{}
+		}
+	}
+
+	uniqueMethods := make([]string, 0)
+	for m := range httpMethods {
+		uniqueMethods = append(uniqueMethods, m)
+	}
+
+	return uniqueMethods
+}
+
+func (p *Plugin) GetRequiredMethods(options []av_opts.AtlasValidateFieldOption_Operation) []string {
+	requiredMethods := make(map[string]struct{}, 0)
+	for _, op := range options {
+		switch op {
+		case av_opts.AtlasValidateFieldOption_create:
+			requiredMethods["POST"] = struct{}{}
+		case av_opts.AtlasValidateFieldOption_update:
+			requiredMethods["PATCH"] = struct{}{}
+			requiredMethods["PUT"] = struct{}{}
+		}
+	}
+
+	uniqueMethods := make([]string, 0)
+	for m := range requiredMethods {
+		uniqueMethods = append(uniqueMethods, m)
+	}
+
+	return uniqueMethods
+}
+
+func (p *Plugin) generateValidateRequired(md *descriptor.DescriptorProto, t string) {
+	requiredFields := make(map[string][]string)
+	for _, fd := range md.GetField() {
+		if fExt, err := proto.GetExtension(fd.Options, av_opts.E_Field); err == nil && fExt != nil {
+			favOpt := fExt.(*av_opts.AtlasValidateFieldOption)
+			methods := p.GetRequiredMethods(favOpt.GetRequired())
+			if len(methods) == 0 {
+				continue
+			}
+			requiredFields[fd.GetName()] = methods
+		}
+	}
+
+	p.P(fmt.Sprintf(`func ValidateRequired_Object_%s(ctx context.Context, v map[string]json.RawMessage) error {`, t))
+	p.P(`method := validate_runtime.HTTPMethodFromContext(ctx)`)
+	p.P(`_ = method`)
+
+	for fn, methods := range requiredFields {
+		if len(methods) == 3 {
+			p.P(fmt.Sprintf(`if _, ok := v["%s"]; !ok {`, fn))
+			p.P(fmt.Sprintf(`return fmt.Errorf("field %s required for %s" )`, fn, strings.Join(methods, ", ")))
+			p.P(`}`)
+		} else {
+			p.P(fmt.Sprintf(`if _, ok := v["%s"]; !ok && (method == "%s"){`, fn, strings.Join(methods, `" || method == "`)))
+			p.P(fmt.Sprintf(`return fmt.Errorf("field %s required for %s" )`, fn, strings.Join(methods, ", ")))
```

Comment:
The error may be misleading. suppose we define required: [create, update]. end-user runs PUT command that triggers an error and receives error message that field 'X' is required for PUT, PATCH. What does it mean? Server does not implement PATCH method but requires 'X' for PATCH. I propose to generate errors in ad-hoc style, saying that "field %s is required for update" and "field %s is required for create". Or, if function call is a case, we should return at least the same method (PUT) that was called instead of giving unrelated information to a user.

Reply:
Sounds good, I will think about it

Reply:
so, just return fmt.Errorf("field %s required for %s", fn, method)